src/test/datascience/notebook/helper.ts

```diff
@@ -287,89 +305,88 @@ export async function waitForKernelToChange(criteria: { labelOrId?: string; inte
         traceInfo(`Active kernel is id:label = ${selectedController.id}:${selectedController.label}`);
         return false;
     };
-    await waitForCondition(
-        async () => isRightKernel(),
-        defaultNotebookTestTimeout,
-        `Kernel with criteria ${JSON.stringify(criteria)} not selected`
-    );
-
-    // Make sure the kernel is actually in use before returning (switching is async)
-    await sleep(500);
+    if (!isRightKernel()) {
+        let tryCount = 0;
+        await waitForCondition(
+            async () => {
+                // Double check not the right kernel (don't select again if already found to be correct)
+                if (!isRightKernel()) {
+                    traceInfoIf(isCI, `Notebook select.kernel command switching to kernel id ${id}: Try ${tryCount}`);
+                    // Send a select kernel on the active notebook editor. Keep sending it if it fails.
+                    await commands.executeCommand('notebook.selectKernel', { id, extension: JVSC_EXTENSION_ID });
```

Comment:
I dont think this needs to be ezecuted multiple times, i think once is sufficient, and after that we just need to keep waiting until the right kernel is selected.
Else we keep firing this command multkiple times & i'm concerned we could have other side effects of filing this unnecessarily

Reply:
That's not what I found. There was a potential race between 'createNotebookController (and results getting to VS code proper)' and the 'notebook.selectKernel' command coming in. This race went away when I did this in a loop.

Reply:
Previously we didn't even care if it worked (hence the comment in the ```setAsActiveControllerForTests```). And we'd set and forget it. But sometimes it didn't work.

With this loop it seems to always work.